File: ChessAndroid/app/src/main/cpp/chess/data/Piece.cpp
#include "Piece.h"

#include "../algorithm/MoveGen.h"

const Piece Piece::EMPTY{};

U64 Piece::getPossibleMoves(const byte square, const Board &board) const noexcept
{
	const Color c = color();
	U64 result{};

	switch (type())
	{
	case PieceType::PAWN:
		result = MoveGen<ALL>::getPawnMoves(c, square, board);
		break;
	case PieceType::KNIGHT:
		result = MoveGen<ALL>::getKnightMoves(c, square, board);
		break;
	case PieceType::BISHOP:
		result = MoveGen<ALL>::getBishopMoves(c, square, board);
		break;
	case PieceType::ROOK:
		result = MoveGen<ALL>::getRookMoves(c, square, board);
		break;
	case PieceType::QUEEN:
		result = MoveGen<ALL>::getQueenMoves(c, square, board);
		break;
	case PieceType::KING:
		result = MoveGen<ALL>::getKingMoves(c, square, board);
		break;
	default:
		break;
	}
	
	return result;
}

U64 Piece::getPossibleCaptures(const byte square, const Board &board) const noexcept
{
	const Color c = color();
	U64 result{};

	switch (type())
	{
	case PieceType::PAWN:
		result = MoveGen<CAPTURES>::getPawnMoves(c, square, board);
		break;
	case PieceType::KNIGHT:
		result = MoveGen<CAPTURES>::getKnightMoves(c, square, board);
		break;
	case PieceType::BISHOP:
		result = MoveGen<CAPTURES>::getBishopMoves(c, square, board);
		break;
	case PieceType::ROOK:
		result = MoveGen<CAPTURES>::getRookMoves(c, square, board);
		break;
	case PieceType::QUEEN:
		result = MoveGen<CAPTURES>::getQueenMoves(c, square, board);
		break;
	case PieceType::KING:
		result = MoveGen<CAPTURES>::getKingMoves(c, square, board);
		break;
	default:
		break;
	}

	return result;
}

File: ChessAndroid/app/src/main/java/net/theluckycoder/chess/model/Piece.kt
package net.theluckycoder.chess.model

class IndexedPiece(
    val id: Int,
    private val square: Int,
    private val type: Byte,
    private val isWhite: Boolean
) {
    fun toPiece() = Piece(square, type, isWhite)
}

data class Piece(
    val square: Int,
    val type: Byte,
    val isWhite: Boolean
) {

    fun getScore() = when (type) {
        PAWN -> 1 // Pawn
        KNIGHT -> 3 // Knight
        BISHOP -> 3 // Bishop
        ROOK -> 5 // Rook
        QUEEN -> 9 // Queen
        else -> 0
    }

    companion object {
        const val PAWN: Byte = 1
        const val KNIGHT: Byte = 2
        const val BISHOP: Byte = 3
        const val ROOK: Byte = 4
        const val QUEEN: Byte = 5
        const val KING: Byte = 6
    }
}
